src/Shared/Buffers.MemoryPool/MemoryPoolBlock.cs

```diff
@@ -6,52 +6,29 @@
 namespace System.Buffers
 {
     /// <summary>
-    /// Block tracking object used by the byte buffer memory pool. A slab is a large allocation which is divided into smaller blocks. The
-    /// individual blocks are then treated as independent array segments.
+    /// Wraps an array allocated in the pinned object heap in a reusable block of managed memory
     /// </summary>
     internal sealed class MemoryPoolBlock : IMemoryOwner<byte>
     {
-        private readonly int _offset;
-        private readonly int _length;
-
-        /// <summary>
-        /// This object cannot be instantiated outside of the static Create method
-        /// </summary>
-        internal MemoryPoolBlock(SlabMemoryPool pool, MemoryPoolSlab slab, int offset, int length)
+        internal MemoryPoolBlock(SlabMemoryPool pool, int length)
         {
-            _offset = offset;
-            _length = length;
-
             Pool = pool;
-            Slab = slab;
 
-            Memory = MemoryMarshal.CreateFromPinnedArray(slab.PinnedArray, _offset, _length);
+            var pinnedArray = GC.AllocateUninitializedArray<byte>(length, pinned: true);
```

Comment:
https://docs.microsoft.com/en-us/dotnet/api/system.gc.allocateuninitializedarray?view=net-5.0 as an FYI to others reviewing.

We didn't zero-allocate the array before, correct?

Reply:
Correct

Reply:
https://github.com/dotnet/aspnetcore/pull/30732/files#diff-e7ec344376781603909872c55a122a48d16c5d9094284d51be214bd5d092fdfbL33